test/hotspot/gtest/gc/z/test_zForwarding.cpp

```diff
@@ -39,6 +42,42 @@ using namespace testing;
 
 class ZForwardingTest : public Test {
 public:
+  // Setup and tear down
+  ZHeap*       _old_heap;
+  ZGenerationOld* _old_old;
+  ZGenerationYoung* _old_young;
+
+  virtual void SetUp() {
+    ZGlobalsPointers::initialize();
+    _old_heap = ZHeap::_heap;
+    ZHeap::_heap = (ZHeap*)os::malloc(sizeof(ZHeap), mtTest);
+
+    _old_old = ZGeneration::_old;
+    _old_young = ZGeneration::_young;
+
+    ZGeneration::_old = &ZHeap::_heap->_old;
+    ZGeneration::_young = &ZHeap::_heap->_young;
+
+    *const_cast<ZGenerationId*>(&ZGeneration::_old->_id) = ZGenerationId::old;
+    *const_cast<ZGenerationId*>(&ZGeneration::_young->_id) = ZGenerationId::young;
+
+    ZGeneration::_old->_seqnum = 1;
+    ZGeneration::_young->_seqnum = 2;
+
+    bool reserved = os::attempt_reserve_memory_at((char*)ZAddressHeapBase, ZGranuleSize, false /* executable */);
+    ASSERT_TRUE(reserved);
```

Comment:
Hi, Thanks for the great work!
I have performed some tests on linux-riscv64 Hifive Unmatched board. So far, I only witnessed one gtest failure:

```
$ make test TEST=gtest:ZForwardingTest
Building target 'test' in configuration 'linux-riscv64-server-release'
Test selection 'gtest:ZForwardingTest', will run:
* gtest:ZForwardingTest/server

Running test 'gtest:ZForwardingTest/server'
Note: Google Test filter = ZForwardingTest*
[==========] Running 4 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 4 tests from ZForwardingTest
[ RUN      ] ZForwardingTest.setup_vm
test/hotspot/gtest/gc/z/test_zForwarding.cpp:68: Failure
Value of: reserved
  Actual: false
Expected: true
[  FAILED  ] ZForwardingTest.setup_vm (0 ms)
[ RUN      ] ZForwardingTest.find_empty_vm
[       OK ] ZForwardingTest.find_empty_vm (1 ms)
[ RUN      ] ZForwardingTest.find_full_vm
[       OK ] ZForwardingTest.find_full_vm (8 ms)
[ RUN      ] ZForwardingTest.find_every_other_vm
[       OK ] ZForwardingTest.find_every_other_vm (0 ms)
[----------] 4 tests from ZForwardingTest (761 ms total)

[----------] Global test environment tear-down
ERROR: RUN_ALL_TESTS() failed. Error 1
[==========] 4 tests from 1 test suite ran. (762 ms total)
[  PASSED  ] 3 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] ZForwardingTest.setup_vm

 1 FAILED TEST
Finished running test 'gtest:ZForwardingTest/server'
Test report is stored in build/linux-riscv64-server-release/test-results/gtest_ZForwardingTest_server

==============================
Test summary
==============================
   TEST                                              TOTAL  PASS  FAIL ERROR
>> gtest:ZForwardingTest/server                          4     3     1     0 <<
==============================
TEST FAILURE
```

The gtest failed this assertion where 'reserved' return by function os::attempt_reserve_memory_at is false.
I find the reason is that the mmap call at the bottom returns a different address instead of the requested one (ZAddressHeapBase). I think that is possible since we are not sure if the requested address is available before the mmap call, right? So I guess we might need some changes here for this gtest.

Reply:
Thanks for reporting. It would be interesting to see what address you get and compare it to the range [ZAddressHeapBase, ZAddressHeapBase+ZAddressOffsetMax).

Reply:
We emailed to erik to discuss this issue two months ago, and maybe he missed it.
ZForwardingTest does not guarantee a successful invoke of os::commit_memory for ZAddressHeapBase, and we saw some conflicts between ZAddressHeapBase and the metadata address space on the RISC-V hardware of 39-bits virtual address. There is no failure in the normal initialization phase of JVM, because the commit order of them is guaranteed.

Reply:
Could you provide the values for `reserved`, `ZAddressHeapBase`, and `ZAddressOffsetMax, when this test is failing. I'd like to know if we can make a workaround for you, or if we have to turn off the test for riscv.

Reply:
@stefank : I ran this gtest for 5 times and here is what I got.
```
ZAddressHeapBase :     0x800000000
ZAddressOffsetMax:     0x800000000
ZGranuleSize          :     0x200000
```
In os::pd_attempt_reserve_memory_at() which is called by os::attempt_reserve_memory_at(), return value by anon_mmap() [1] is one of: 0x3f8d5ff000, 0x3f649fe000, 0x3f5d3ff000, 0x3f68077000 and 0x3f555ff000

So seems that those values are not in the range [ZAddressHeapBase, ZAddressHeapBase+ZAddressOffsetMax).

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/os/linux/os_linux.cpp#L3334

Reply:
That's unfortunate. Could you try this patch, which probes the address range to see if it can reserve the memory somewhere else within `[ZAddressHeapBase, ZAddressHeapBase+ZAddressOffsetMax)`:
https://github.com/stefank/jdk/tree/zgc_generational_review_test_zforwarding

Reply:
> That's unfortunate. Could you try this patch, which probes the address range to see if it can reserve the memory somewhere else within `[ZAddressHeapBase, ZAddressHeapBase+ZAddressOffsetMax)`: https://github.com/stefank/jdk/tree/zgc_generational_review_test_zforwarding

@stefank : Good news is that this gtest case can now pass on linux-riscv64 with this patch. I tried several times and it seems we could always reserve a space of ZGranuleSize bytes at address 0x852000000. Thanks for fixing this!

Reply:
Thanks for helping out with this. I've now pushed the proposed patch.